### PR TITLE
#48 - Auth service layer integration with login screen

### LINKED ISSUE #48: Create a service layer
## Description:
Implement the service layer by integrating the safient core authentication flow.

### LINKED ISSUE #76: API integration - Login screen

### PR DESCRIPTION
**Pull request by devpavan04**
## Description
Implements the service layer by integrating safient core authentication flow. The user information is then populated on the account page.

- [x] Account service integration with Safient Core SDK
- [x] Login screen and account screen integration with auth service

## Screenshot
https://www.loom.com/share/f744754e01e24a16a50a5d8b7e50aa1f

## Platform tested
- [x] Web
- [x] Tablet
- [x] Mobile

### CODE REVIEW

File: src/store/account/account.store.impl.ts [hunk 1]
```diff
@@ -34,31 +50,44 @@ export class AccountStoreImpl extends StoreImpl implements AccountStore {
     storageService.remove(StorageKey.userId);
   }
 
+  async initializeSafient(web3Provider: Web3Provider): Promise<void> {
```
Comment (koshikraj) [inline src/store/account/account.store.impl.ts]: This should be moved to service, right?

File: src/screens/account/account.screen.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                )}
+              </>
+     
+        
+      </Card>
+    </AccountScreenContainer>
+  );
+});
+
+window.ethereum &&
```
Comment (dsouzaedison) [inline src/screens/account/account.screen.tsx]: Create a new service called ethereum service and move these listeners there.

File: src/screens/auth/login/login.screen.tsx [hunk 3]
```diff
@@ -14,6 +18,12 @@ import {
 } from './login.screen.styles';
 
 export const LoginScreen = () => {
+
+  const { accountService } = useServices();
+  let history = useHistory();  
+
+  const login = async () => {const account = await accountService.login(); if (account.hasData()) history.push(RoutePath.home)};
```
Comment (dsouzaedison) [inline src/screens/auth/login/login.screen.tsx]: Please prettify this line

File: src/services/account/account.service.impl.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  constructor() {
+    super();
+    this.accountStore = stores?.accountStore;
+    this.web3Modal = new Web3Modal({
+      cacheProvider: true,
+      theme: 'light',
+    });
+  }
+
+  async login(): Promise<ServiceResponse<Types.User | undefined>> {
```
Comment (dsouzaedison) [inline src/services/account/account.service.impl.ts]: Is undefined required in `ServiceResponse<Types.User | undefined>` ?

File: src/screens/account/account.screen.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            <Text variant='contentHeader' text='Address' />
+            <Text variant='content' text={stores.accountStore.address} />
+            <br />
+            <Text variant='contentHeader' text='Balance' />
+            <Text variant='content' text={stores.accountStore.balance} />
+            <br />
+            <br />
+            <br />
+            <br />
+
```
Comment (dsouzaedison) [inline src/screens/account/account.screen.tsx]: Using so many break tags is not right. Make use of CSS for spacing.

**Review by dsouzaedison — CHANGES_REQUESTED** [review]

File: src/screens/account/account.screen.tsx (hunk 5 above)
Comment (koshikraj) [reply src/screens/account/account.screen.tsx]: Yeah, sure. This page content was a placeholder. We shall update.

File: src/services/account/account.service.impl.ts (hunk 4 above)
Comment (koshikraj) [reply src/services/account/account.service.impl.ts]: I think we can eliminate this.

File: src/services/core/service-response..ts [hunk 6]
```diff
@@ -1,8 +1,8 @@
 export class ServiceResponse<T> {
   readonly data?: T;
-  private readonly error?: Error;
+  readonly error?: T;
 
-  constructor({ data, error }: { data?: T; error?: Error }) {
+  constructor({ data, error }: { data?: T; error?: T }) {
```
Comment (dsouzaedison) [inline src/services/core/service-response..ts]: Error cannot be of type T. T represents valid response.

File: src/services/ethereum/ethereum.service.impl.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  constructor() {
+    super();
+    this.listen()
+    // this.accountStore = stores?.accountStore;
+   
+  }
+
+  async listen(): Promise<void> {
+    
+    window.ethereum &&
```
Comment (dsouzaedison) [inline src/services/ethereum/ethereum.service.impl.ts]: Use if statements for non-JSX code and prettify this file.

File: src/services/ethereum/ethereum.service.impl.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    setTimeout(() => {
+      //  accountService.login();
+    }, 1);
+  });
+
+window.ethereum &&
+  window.ethereum.on('accountsChanged', () => {
+    console.log('change chain')
+    setTimeout(() => {
+      // accountService.login();
```
Comment (dsouzaedison) [inline src/services/ethereum/ethereum.service.impl.ts]: Is this setTimeout required?

File: src/store/account/account.store.impl.ts [hunk 9]
```diff
@@ -55,14 +55,13 @@ export class AccountStoreImpl extends StoreImpl implements AccountStore {
     this.safient = new SafientCore(
       this.signer,
       Enums.NetworkType.localhost,
-      'threadDB',
+      Enums.DatabaseType.threadDB,
```
Comment (dsouzaedison) [inline src/store/account/account.store.impl.ts:58]: There's no need to bind enums inside Enums. `DatabaseType.threadDB` would do. Enums declarations must go into respective folders based on their context.

**Review by dsouzaedison — CHANGES_REQUESTED** [review]

File: src/screens/account/account.screen.tsx [hunk 10]
```diff
@@ -1,53 +1,93 @@
...
+  DefaultSettingsContainer,
+  Card,
+  StyledInput,
+  InputContainer,
+  SwitchContainer,
+} from './account.screen.styles';
+// Todo- remove this after integration
 import { stores } from '../../store';
-import { AccountScreenContainer, Title } from './account.screen.styles';
+import Demo from '../../../src/assets/images/demo.png';
```
Comment (dsouzaedison) [inline src/screens/account/account.screen.tsx]: Image imports should go inside image.ts. src prop must be used only for dynamic images.

**Review by dsouzaedison — CHANGES_REQUESTED** [review]

**Review by koshikraj — APPROVED** [review]

**Pull request merged**